[operator] Reconcile `seedmanagement.gardener.cloud/v1alpha1.Gardenlet` object when there is no corresponding `Seed` yet (#10161)
gardener/gardener: docs/deployment/deploy_gardenlet_via_operator.md

# Deploy a gardenlet Via `gardener-operator`

The gardenlet can automatically be deployed by `gardener-operator` into existing Kubernetes clusters in order to register them as seeds.

## Prerequisites

Using this method only works when [`gardener-operator`](../concepts/operator.md) is managing the garden cluster.
If you have used the [`gardener/controlplane` Helm chart](../../charts/gardener/controlplane) for the deployment of the Gardener control plane, please refer to [this document](deploy_gardenlet_manually.md).

> [!TIP]
> The initial seed cluster can be the garden cluster itself, but for better separation of concerns, it is recommended to only register other clusters as seeds.

## Deployment of gardenlets

Using this method, `gardener-operator` is only taking care of the very first deployment of gardenlet.

> Once running, the gardenlet leverage the [self upgrade](deploy_gardenlet_manually.md#self-upgrades) strategy in order to keep themselves up-to-date.

Reviewer: ```suggestion
Once running, the gardenlet leverages the [self upgrade](deploy_gardenlet_manually.md#self-upgrades) strategy in order to keep itself up-to-date.
```

Alternatively use `gardenlet`s instead.

    kind: GardenletConfiguration
    controllers:
      shoot:
        reconcileInMaintenanceOnly: true
        respectSyncPeriodOverwrite: true
      shootState:
        concurrentSyncs: 0
    featureGates:
      ShootManagedIssuer: true
    etcdConfig:
      featureGates:
        UseEtcdWrapper: true
    logging:
      enabled: true
      vali:
        enabled: true
      shootNodeLogging:
        shootPurposes:
        - infrastructure
        - production
        - development
        - evaluation
    seedConfig:
      apiVersion: core.gardener.cloud/v1beta1
      kind: Seed
      metadata:
        labels:
          base: kind
      spec:
        backup:
          provider: local
          region: local
          secretRef:
            name: backup-local
            namespace: garden
        dns:
          provider:
            secretRef:
              name: internal-domain-internal-local-gardener-cloud
              namespace: garden
            type: local
        ingress:
          controller:
            kind: nginx
          domain: ingress.local.seed.local.gardener.cloud
        networks:
          nodes: 172.18.0.0/16
          pods: 10.1.0.0/16

>           services: 10.2.0.0/16

Reviewer: Is this required if deploying into the garden runtime cluster? Does `gardener-operator` not have that knowledge already? In other words, can we reduce the amount of configuration necessary for the "simple" case of garden runtime cluster = first seed cluster?

Author: Maybe, but this is something that we should probably do separately. `gardener-operator` would need a webhook on `Seed`s and default things that are already known by the `Garden` resource. However, it's a very neat/special case, so not sure if it's generally worth the effort/code/complexity.

          shootDefaults:
            pods: 10.3.0.0/16
            services: 10.4.0.0/16
        provider:
          region: local
          type: local
          zones:
          - "0"
        settings:
          excessCapacityReservation:
            enabled: false
          scheduling:
            visible: true
          verticalPodAutoscaler:
            enabled: true
```

This causes `gardener-operator` to deploy gardenlet to the same cluster where it is running.

> Once it came up, gardenlet will create a `Seed` resource with the same name and uses the `Gardenlet` resource for self-upgrades (see [this document](deploy_gardenlet_manually.md#self-upgrades)).

Reviewer: ```suggestion
Once it comes up, gardenlet will create a `Seed` resource with the same name and uses the `Gardenlet` resource for self-upgrades (see [this document](deploy_gardenlet_manually.md#self-upgrades)).
```

### Remote Clusters

> If you want it to deploy gardenlet into some other cluster, create a kubeconfig `Secret` and reference it in the `Gardenlet` resource:

Reviewer: ```suggestion
If you want `gardener-operator` to deploy gardenlet into some other cluster, create a kubeconfig `Secret` and reference it in the `Gardenlet` resource:
```
